### ISSUE TITLE
Improve the performance of ed25519

### ISSUE THREAD

**Issue by zmanian** [tendermint/tendermint#1923]
@gtank's optimized assembly implementation of ed25519 provides a 50% speed up on signature verification on amd64.
https://go-review.googlesource.com/c/crypto/+/71950

Let's break this out into `/crypto` and use it for verification.

Then lets add a batch verification algorithm that only is used during fast sync. (ref #1809 )

**Comment by xla** [comment]
Can this be done post-launch? Or has this to be proven to be the last remaining bottleneck for fast sync?

**Comment by zmanian** [comment]
This is 100% post launch IMO

**Comment by zmanian** [comment]
but @ValarDragon can start a branch for this if he wants

**Comment by hdevalence** [comment]
I believe that the code referenced above has been folded into this repo:

https://github.com/gtank/ristretto255

The internal edwards25519 package exposes fast group operations, and in particular has a fast implemnetatino of multiscalar multiplication.  This means that it can be used for implementing batch verification (#1809), but it's worth checking that batch verification won't affect the consensus logic.

**Comment by zmanian** [comment]
The model we've been considering is to only do batch verification during the catch up phase "fast sync"

**Comment by hdevalence** [comment]
Cool, that makes sense. I'm kind of curious how many signatures are in a block and whether it's possible to batch across blocks.

**Comment by zmanian** [comment]
100 signatures per block in consensus on the Hub.

Batching across blocks would probably have a much bigger impact on blockchain reactor control flow. In principle, it's possible. It's incredibly unlikely that a signature verification would fail during fast sync.

**Comment by marbar3778** [comment]
closed by https://github.com/tendermint/tendermint/pull/5760

**Comment by ethanfrey** [comment]
Very nice.

Did you use the batch verification in the #5760?
The text says version bump and I read a request to help implement batch verification a week or so ago.

2x speed is great. Just wondering if more if possible

**Comment by ethanfrey** [comment]
Ah, saw that https://github.com/hdevalence/ed25519consensus/issues/5 is still open, but that is tracked by https://github.com/tendermint/tendermint/issues/1809

**Comment by marbar3778** [comment]
> Very nice.
> 
> Did you use the batch verification in the #5760?
> The text says version bump and I read a request to help implement batch verification a week or so ago.
> 
> 2x speed is great. Just wondering if more if possible

Batch verification was not added. There is an open issue to add it (https://github.com/hdevalence/ed25519consensus/issues/5) but no one has come forward yet. We will need to rework verification once it lands.
